### PR TITLE
refactor: replace $stacked-variables with accumulator

### PR DESCRIPTION
**Pull request by AirQuick**
Local variables are stacked using a tunnel parameter (`$stacked-variables`) in unusual sibling-walking templates.

To simplify the logic, this pull request replaces the tunnel parameter with an accumulator (`stacked-variables-distinct-uqnames`).

### CODE REVIEW

File: src/compiler/base/main.xsl [hunk 1]
```diff
@@ -27,6 +28,32 @@
...
       select="x:document-actual-uri($initial-document)" />
 
+   <!--
+      Accumulators for non-global x:variable
+   -->
+
+   <!-- Push and pop x:variable based on node identity -->
+   <xsl:accumulator name="local:stacked-variables" as="element(x:variable)*" initial-value="()">
+      <xsl:accumulator-rule match="x:scenario/x:variable" select="$value, ." />
+      <xsl:accumulator-rule match="x:scenario" phase="end" select="$value except x:variable" />
```
Comment (galtm) [inline src/compiler/base/main.xsl:38]: Initially, I misread the `select` attribute, forgetting that `x:variable` is relative to `x:scenario`. Your code is correct, but I wonder if something like `$value except ./child::x:variable` or `<!-- Remove variables declared as children of this scenario -->` would make it a bit harder for that kind of misreading to happen. I haven't worked with accumulators much, which might have contributed to my misunderstanding, so it's fine if you prefer to leave this as is.

**Review by galtm — APPROVED** [review]
The diffs look good, and when I exercised the code, it worked as expected.

File: src/compiler/base/main.xsl (hunk 1 above)
Comment (AirQuick) [reply src/compiler/base/main.xsl]: Added comments and explicit axes to the expressions. (958ead095d433d32b2603cf037871d7aa6dd5d9d)

**Comment by galtm** [comment]
@AirQuick , do you think the compiler's `pending` tunnel parameter could also be replaced with an accumulator? If so, do you think it would be a worthwhile change? Things I liked about your solution in this PR are that the accumulator rules are in one place and understanding the usage doesn't require you to examine call stacks to determine which caller might have passed a tunnel parameter.

**Comment by AirQuick** [comment]
It would be a worthwhile change. Right now the tunnel `$pending`, `ancestor::x:scenario/@pending`, `ancestor::*/@focus` and `$pending-p` (what is this `-p`?) are tangled in a way I can't figure out instantly. Resolving them in a single place would be a great step forward.
Added a comment: https://github.com/xspec/xspec/issues/575#issuecomment-697206772

**Comment by galtm** [comment]
> Added a comment: [#575 (comment)](https://github.com/xspec/xspec/issues/575#issuecomment-697206772)

Thanks, that's a reasonable place to record the idea for future reference.

My guess is that whoever introduced the "-p" at the end of the variable name was familiar with conventions like http://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node69.html and just meant "is this pending, yes or no?"

**Comment by AirQuick** [comment]
> My guess is that whoever introduced the "-p" at the end of the variable name was familiar with conventions like http://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node69.html and just meant "is this pending, yes or no?"

Thanks! Now I get it.
Some lines in the XSpec code base refer to "predicate" in a context broader than XPath's. Probably the same mind-set.

**Pull request merged**